### PR TITLE
✨ feat(gentest): Migrate `gentest` to Jinja2

### LINKED ISSUE #863: 🔀(gentest): use jinja2 templates to structure python output
[Jinja2](https://jinja.palletsprojects.com/en/3.1.x/) is a very powerful templating library that we started using in the doc flow for the generated "Test Case Reference" output pages, see #801 and `./docs/templates`.

I think that `.py.j2` templates could be an excellent way to write the output from `gentest`. This will be particularly helpful, when/if we create `blockchain_tests` with an arbitrary number of blocks, each containing an arbitrary number of transactions. We could have a template for a `state_test` and one for `blockchain_test`. Jinja2 also allows base templates.

### PR DESCRIPTION
**Pull request by raxhvl**
## 🗒️ Description

This PR introduces the use of Jinja2 templates for generating test scripts, contained within a new `gentest` package that includes self-contained templates and related modules.

```console
src/cli/gentest
├── 📁 templates/blockchain_test  [Jinja2 templates]
│   └── transaction.py.j2 [Template for creating transaction based gentest]
├── 📄 __init__.py
├── 📄 request_manager.py [A request manager Ethereum  RPC calls]
├── 📄 cli.py [CLI interface for generating blockchain test scripts]
└── 📄 test_providers.py  [Context providers for creating test scripts]
```
## Notes on code refactoring 

The code has been refactored for simplicity. For instance, the previous implementation used a `RemoteTransaction` dataclass:

```python
@dataclass()
class RemoteTransaction:
    """
    Remote transaction structure
    """

    block_number: int
    tr_hash: Hash
    transaction: Transaction
```

This required nested calls like `tr.transaction`: https://github.com/ethereum/execution-spec-tests/blob/6fb195be140a1d7ce450f7a083032f56828928b8/src/cli/gentest.py#L128

The class has been flattened by making it inherit from the `Transaction` type, which now provides Pydantic validations out of the box: 
```python
    class RemoteTransaction(Transaction):
        """
        Model that represents a transaction.
        """

        block_number: HexNumber
        tx_hash: Hash
```
[source](https://github.com/raxhvl/execution-spec-tests/blob/d86b4319733b32b45a94467a40c26c007a02bc4c/src/cli/gentest/request_manager.py#L28-L34.)

This results in a simpler interface: 

```python
if address == self.transaction.sender:
```
[source](https://github.com/raxhvl/execution-spec-tests/blob/d86b4319733b32b45a94467a40c26c007a02bc4c/src/cli/gentest/test_providers.py#L62)

More importantly, pydantic rightly identified that `to` field of transaction can be null which is now handled: 
```python
assert transaction.to is not None
```
[source](https://github.com/raxhvl/execution-spec-tests/blob/fca443dc0e9680866a2a84dad3120d25608c4e15/src/cli/gentest/request_manager.py#L107)

The `test_providers` module provides a predictable interface that provides context for the templates and is designed to play nice with unit testing.

## Future Improvements
The generated tests can be further improved later. This PR focuses solely on the migration process, allowing for testing by diffing the generated tests against those from the main branch.

## 🔗 Related Issues
closes #863

## ✅ Checklist
- [x] All: Set appropriate labels for the changes.
- [x] All: Considered squashing commits to improve commit history.
- [x] All: Added an entry to [CHANGELOG.md](/ethereum/execution-spec-tests/blob/main/docs/CHANGELOG.md).
- [x] All: Considered updating the online docs in the [./docs/](/ethereum/execution-spec-tests/blob/main/docs/) directory.
- [x] Tests: All converted JSON/YML tests from [ethereum/tests](/ethereum/tests) have been added to [converted-ethereum-tests.txt](/ethereum/execution-spec-tests/blob/main/converted-ethereum-tests.txt).
- [x] Tests: A PR with removal of converted JSON/YML tests from [ethereum/tests](/ethereum/tests) have been opened.
- [x] Tests: Included the type and version of evm t8n tool used to locally execute test cases:  e.g., ref with commit hash or geth 1.13.1-stable-3f40e65.
- [x] Tests: Ran `mkdocs serve` locally and verified the auto-generated docs for new tests in the [Test Case Reference](https://ethereum.github.io/execution-spec-tests/main/tests/) are correctly formatted.

### CODE REVIEW

**Review by danceratopz — COMMENTED** [review]
Thanks! Looks like a great start to adding jinja2 templating! 

I know it wasn't the focus of your PR, but I couldn't help but try and improve the output a little. The suggestions below should ensure that the generated output passes flake8/isort/mypy checks, but I might have missed something :)

File: src/cli/gentest/request_manager.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+from pydantic import BaseModel
+
+from config import EnvConfig
+from ethereum_test_base_types import Account, Address, Hash, HexNumber
+from ethereum_test_rpc import BlockNumberType, DebugRPC, EthRPC
+from ethereum_test_types import Transaction
+
+
+class RequestManager:
```
[the remark this answers is not in the document]
Comment (raxhvl) [reply src/cli/gentest/request_manager.py:23]: >previously struggled with the name
I feel seen.

**Comment by raxhvl** [comment]
I passed `tx_hash` to the template. This gets rid of the verbose methods like `_get_module_docstring`, `_get_test_name`, and `_get_test_docstring`. Generally, whenever I see long strings or `\n` in code, I take it as an indication that the Python code has borrowed some templating logic. The TODOs I left follow this reasoning, and I will revisit them at some point.

**Comment by danceratopz** [comment]
> The TODOs I left follow this reasoning, and I will revisit them at some point.

Got it!

**Review by danceratopz — APPROVED** [review]
Thanks for including the suggestions!

**Pull request merged**